File: developers/academy/py/named_vectors/_snippets/103_searches.py
# START-ANY
import weaviate
import weaviate.classes.query as wq
import os

# END-ANY

# FilteredSemanticSearch
from datetime import datetime

# END FilteredSemanticSearch

# START-ANY

# END-ANY

headers = {"X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")}
client = weaviate.connect_to_local(
    port=8280,
    grpc_port=50251,
    headers=headers
)

# START-ANY
# Instantiate your client (not shown). e.g.:
# headers = {"X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY")}  # Replace with your OpenAI API key
# client = weaviate.connect_to_local(headers=headers)

# END-ANY


# NVTitleSearch
# Get the collection
movies = client.collections.get("MovieNVDemo")

# Perform query
response = movies.query.near_text(
    query="A joyful holiday film",
    # highlight-start
    target_vector="title",  # The target vector to search against
    # highlight-end
    limit=5,
    return_metadata=wq.MetadataQuery(distance=True),
    return_properties=["title", "release_date", "tmdb_id", "poster"]
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year, o.properties["tmdb_id"]
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END NVTitleSearch


print("\n\n")

client.connect()


# NVOverviewSearch
# Get the collection
movies = client.collections.get("MovieNVDemo")

# Perform query
response = movies.query.near_text(
    query="A joyful holiday film",
    # highlight-start
    target_vector="overview",  # The target vector to search against
    # highlight-end
    limit=5,
    return_metadata=wq.MetadataQuery(distance=True),
    return_properties=["title", "release_date", "tmdb_id", "poster"]
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year, o.properties["tmdb_id"]
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END NVOverviewSearch


print("\n\n")

client.connect()


# NVPosterSearch
# Get the collection
movies = client.collections.get("MovieNVDemo")

# Perform query
response = movies.query.near_text(
    query="A joyful holiday film",
    # highlight-start
    target_vector="poster_title",  # The target vector to search against
    # highlight-end
    limit=5,
    return_metadata=wq.MetadataQuery(distance=True),
    return_properties=["title", "release_date", "tmdb_id", "poster"]
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year, o.properties["tmdb_id"]
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END NVPosterSearch


print("\n\n")

client.connect()


# MetadataMultimodalSearch

def url_to_base64(url):
    import requests
    import base64

    image_response = requests.get(url)
    content = image_response.content
    return base64.b64encode(content).decode("utf-8")


# Get the collection
movies = client.collections.get("MovieNVDemo")

# Perform query
src_img_path = "https://github.com/weaviate-tutorials/edu-datasets/blob/main/img/International_Space_Station_after_undocking_of_STS-132.jpg?raw=true"
query_b64 = url_to_base64(src_img_path)

response = movies.query.near_image(
    near_image=query_b64,
    limit=5,
    target_vector="poster_title",  # The target vector to search against
    return_metadata=wq.MetadataQuery(distance=True),
    return_properties=["title", "release_date", "tmdb_id", "poster"]  # To include the poster property in the response (`blob` properties are not returned by default)
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year, o.properties["tmdb_id"]
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Distance to query: {o.metadata.distance:.3f}\n"
    )  # Print the distance of the object from the query

client.close()
# END MetadataMultimodalSearch


print("\n\n")

client.connect()


# MetadataHybridSearch
# Get the collection
movies = client.collections.get("MovieNVDemo")

# Perform query
response = movies.query.hybrid(
    query="history",
    # highlight-start
    target_vector="overview",  # The target vector to search against
    # highlight-end
    limit=5,
    return_metadata=wq.MetadataQuery(score=True)
)

# Inspect the response
for o in response.objects:
    print(
        o.properties["title"], o.properties["release_date"].year
    )  # Print the title and release year (note the release date is a datetime object)
    print(
        f"Hybrid score: {o.metadata.score:.3f}\n"
    )  # Print the hybrid search score of the object from the query

client.close()
# END MetadataHybridSearch


print("\n\n")

client.connect()

File: developers/academy/js/starter_text_data/_snippets/103_searches.ts
// # START-ANY
import weaviate, { WeaviateClient } from 'weaviate-client'

let client: WeaviateClient
let response

// # END-ANY


client = await weaviate.connectToWeaviateCloud(
    process.env.WCD_URL as string,
    {
        authCredentials: new weaviate.ApiKey(process.env.WCD_API_KEY as string),
        headers: {
            'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY as string,  // Replace with your inference API key
        }
    }
)

// # START-ANY
// Instantiate your client (not shown). e.g.:
// const requestHeaders = {'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY as string,}
// client = weaviate.connectToWeaviateCloud(..., headers: requestHeaders) or
// client = weaviate.connectToLocal(..., headers: requestHeaders)

// # END-ANY

// MetadataSemanticSearch
// Get the collection
// END MetadataSemanticSearch
// MetadataBM25Search // MetadataSemanticSearch // MetadataHybridSearch // FilteredSemanticSearch
const movies = client.collections.get("Movie")
// END MetadataBM25Search // END MetadataSemanticSearch // END MetadataHybridSearch // END FilteredSemanticSearch
// MetadataSemanticSearch

// Perform query
response = await movies.query.nearText('dystopian future', {
    limit: 5,
    returnMetadata: ['distance']
})

// Inspect the response
for (let item of response.objects) {
    // Print the title and release year 
    console.log(`${item.properties.title}: ${item.properties.release_date.getUTCFullYear()} `)
    // Print the distance of the object from the query
    console.log(`Distance to query: ${item.metadata.distance}`)
}
// END MetadataSemanticSearch

// MetadataBM25Search

response = await movies.query.bm25('history', {
    limit: 5,
    returnMetadata: ['score']
})

// Inspect the response
for (let item of response.objects) {
    // Print the title and release year 
    console.log(`${item.properties.title}: ${item.properties.release_date.getUTCFullYear()} `)
    // Print the BM25 score of the object from the query
    console.log(`BM25 score: ${item.metadata.score}`)
}
// END MetadataBM25Search


// MetadataHybridSearch

response = await movies.query.hybrid('history', {
    limit: 5,
    returnMetadata: ['score']
})

// Inspect the response
for (let item of response.objects) {
    // Print the title and release year 

    console.log(`${item.properties.title}: ${item.properties.release_date.getUTCFullYear()} `)
    // Print the hybrid search score of the object from the query

    console.log(`Hybrid score: ${item.metadata.score}`)
}
// END MetadataHybridSearch


// FilteredSemanticSearch

// Perform query
response = await movies.query.nearText('dystopian future', {
    limit: 5,
    returnMetadata: ['distance'],
    // highlight-start
    filters: movies.filter.byProperty('release_date').greaterThan(new Date('December 17, 1995'))
    // highlight-end
})

// Inspect the response
for (let item of response.objects) {
    // Print the title and release year 
    console.log(`${item.properties.title}: ${item.properties.release_date.getUTCFullYear()} `)
    // Print the distance of the object from the query
    console.log(`Distance to query: ${item.metadata.distance}`)
}
client.close()
// END FilteredSemanticSearch
